consensus/polybft/checkpoint_manager.go

```diff
@@ -0,0 +1,180 @@
+		"bytes32 eventRoot, tuple(address _address, uint256[4] blsKey)[] nextValidators" + ")")
+)
+
+type checkpointManager struct {
+	sender           types.Address
+	blockchain       blockchainBackend
+	rootchain        rootchainInteractor
+	consensusBackend polybftBackend
+	epochSize        uint64
+}
+
+// getCurrentCheckpointID queries CheckpointManager smart contract and retrieves current checkpoint id
+func (c checkpointManager) getCurrentCheckpointID(epochNumber uint64) (uint64, error) {
+	checkpointIDMethodEncoded, err := currentCheckpointIDMethod.Encode([]interface{}{})
+	if err != nil {
+		return 0, fmt.Errorf("failed to encode currentCheckpointId function parameters for epoch=%d: %w",
+			epochNumber, err)
+	}
+
+	currentCheckpointID, err := c.rootchain.Call(c.sender, helper.CheckpointManagerAddress, checkpointIDMethodEncoded)
+	if err != nil {
+		return 0, fmt.Errorf("failed to invoke currentCheckpointId function on the rootchain for epoch=%d: %w",
+			epochNumber, err)
+	}
+
+	checkpointID, err := strconv.ParseUint(currentCheckpointID, 0, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to convert current checkpoint id '%s' to number for epoch=%d: %w",
+			currentCheckpointID, epochNumber, err)
+	}
+
+	return checkpointID, nil
+}
+
+// submitCheckpoint sends a transaction which with checkpoint data to the rootchain
+func (c checkpointManager) submitCheckpoint(latestHeader types.Header, epochNumber uint64) error {
+	checkpointID, err := c.getCurrentCheckpointID(epochNumber)
+	if err != nil {
+		return err
+	}
+
+	nonce, err := c.rootchain.GetPendingNonce(c.sender)
+	if err != nil {
+		return err
+	}
+
+	checkpointManagerAddr := ethgo.Address(helper.CheckpointManagerAddress)
+	txn := &ethgo.Transaction{
+		To: &checkpointManagerAddr,
+	}
+
+	// detect any pending (previously failed) checkpoints and send them
+	for blockNum := checkpointID + 1; blockNum < latestHeader.Number; blockNum++ {
+		header, exists := c.blockchain.GetHeaderByNumber(blockNum)
+		if !exists {
+			return fmt.Errorf("block %d was not found", blockNum)
+		}
+
+		extra, err := GetIbftExtra(header.ExtraData)
+		if err != nil {
+			return err
+		}
+
+		// send checkpoints only for epoch ending blocks (validators delta is non-empty)
+		if extra.Validators.IsEmpty() {
+			continue
+		}
+
+		err = c.submitCheckpointInternal(nonce, txn, *header, *extra)
+		if err != nil {
+			return err
+		}
+		nonce++
+	}
+
+	extra, err := GetIbftExtra(latestHeader.ExtraData)
+	if err != nil {
+		return err
+	}
+
+	return c.submitCheckpointInternal(nonce, txn, latestHeader, *extra)
+}
+
+// submitCheckpointInternal encodes checkpoint data for the given block and
+// sends a transaction to the CheckpointManager rootchain contract
+func (c *checkpointManager) submitCheckpointInternal(nonce uint64, txn *ethgo.Transaction,
+	header types.Header, extra Extra) error {
+	nextEpochValidators, err := c.consensusBackend.GetValidators(header.Number, nil)
```

Comment:
If checkpoint is sent on a non-ending epoch block, `nextValidatorSet` would be the same. What did we agree, do we send an empty set or just send the same validator set? @b00ris

Reply:
We have agreed to send the same validator set (https://github.com/0xPolygon/polygon-edge/pull/840#discussion_r1009087586).

Reply:
We agreed to keep to sending  next validator set = current validator for non-end-of-epoch blocks.

Reply:
I think we should we ask to add a wrapper for submitCheckpoint without sending full validator set(send only nextValidatorSetHash).

Reply:
> I think we should we ask to add a wrapper for submitCheckpoint without sending full validator set(send only nextValidatorSetHash).

@b00ris could you elaborate a bit more? `submitCheckpoint` function expects full validator set. Are you talking about adding additional function on smart contract, through which we could provide `nextValidatorSetHash` (although SC should have enough information to calculate it on its own)...?